util: add support to run commands with nsenter (#2983)
ceph/ceph-csi: examples/README.md

Storage class and snapshot class, using `<cluster-id>` as the value for the
option `clusterID`, can now be created on the cluster.

## Running CephCSI with pod networking

The current problem with Pod Networking, is when a CephFS/RBD volume is mounted
in a pod using Ceph CSI and then the CSI CephFS/RBD plugin is restarted or
terminated (e.g. by restarting or deleting its DaemonSet), all operations on
the volume become blocked, even after restarting the CSI pods.

The only workaround is to restart the node where the Ceph CSI plugin pod was
restarted.  This can be mitigated by running the `rbd map`/`mount -t` commands
in a different namespace which does not get deleted and recreated often.

If someone wants to run the CephCSI with the pod networking they can still do
by setting the `netNamespaceFilePath`. If this path is set CephCSI will execute
the `rbd map`/`mount -t` commands by entering the network namespace specified by

> `netNamespaceFilePath` with the nsenter.

Reviewer: > ... with the `nsenter` command.

Or just

> ... with `nsenter`.

Might be good to reference `man 7 namespaces` and/or `man 7 network_namespaces` for additional information.

Author: added link to man page

Storage class and snapshot class, using `<cluster-id>` as the value for the
option `clusterID`, can now be created on the cluster.

## Running CephCSI with pod networking

The current problem with Pod Networking, is when a CephFS/RBD volume is mounted
in a pod using Ceph CSI and then the CSI CephFS/RBD plugin is restarted or
terminated (e.g. by restarting or deleting its DaemonSet), all operations on
the volume become blocked, even after restarting the CSI pods.

The only workaround is to restart the node where the Ceph CSI plugin pod was
restarted.  This can be mitigated by running the `rbd map`/`mount -t` commands

> in a different namespace which does not get deleted and recreated often.

Reviewer 2: The namespace doesn't ever get recreated:
```suggestion
in a different network namespace which does not get deleted when the CSI CephFS/RBD plugin is restarted or terminated.
```

If someone wants to run the CephCSI with the pod networking they can still do
by setting the `netNamespaceFilePath`. If this path is set CephCSI will execute

> the `rbd map`/`mount -t` commands by entering the [network

Reviewer 2: ```suggestion
the `rbd map`/`mount -t` commands after entering the [network
```

namespace](https://man7.org/linux/man-pages/man7/network_namespaces.7.html)
specified by `netNamespaceFilePath` with the
[nsenter](https://man7.org/linux/man-pages/man1/nsenter.1.html) command.

> `netNamespaceFilePath` is typical of the network namespace of a long-running

Reviewer 2: ```suggestion
`netNamespaceFilePath` should point to the network namespace of some long-running
```

> process it usually `/proc/<long running process id>/ns/net`

Reviewer 2: ```suggestion
process, typically it would be a symlink to `/proc/<long running process id>/ns/net`.
```

> The long-running process can also be another pod which is a (Daemonset) which

Reviewer 2: ```suggestion
The long-running process can also be another pod which is a DaemonSet which
```

never restarts. This Pod should only be stopped and restarted when a node is
stopped so that volume operations do not become blocked. The new DaemonSet pod

> can contain a single container, the container responsible for holding
> filesystem mounts and mapped block devices. It is used as a passthrough by the

Reviewer 2: It wouldn't "hold" mounts or block devices -- in fact they don't need holding.  It's all about the pod network:
```suggestion
can contain a single container, responsible for holding its pod network alive.
It is used as a passthrough by the
```

CephCSI plugin pod which when mounting or mapping will use the network
namespace of this pod.

> Once the pod is created get the PID of it and create a symlink of
> `/proc/$$/ns/net` in the hostPath volume shared with the csi-plugin pod and

Reviewer 2: ```suggestion
Once the pod is created get its PID and create a symlink to
`/proc/<PID>/ns/net` in the hostPath volume shared with the csi-plugin pod and
```

by setting the `netNamespaceFilePath`. If this path is set CephCSI will execute
the `rbd map`/`mount -t` commands after entering the [network
namespace](https://man7.org/linux/man-pages/man7/network_namespaces.7.html)
specified by `netNamespaceFilePath` with the
[nsenter](https://man7.org/linux/man-pages/man1/nsenter.1.html) command.

`netNamespaceFilePath` should point to the network namespace of some
long-running process, typically it would be a symlink to
`/proc/<long running process id>/ns/net`.

The long-running process can also be another pod which is a Daemonset which
never restarts. This Pod should only be stopped and restarted when a node is
stopped so that volume operations do not become blocked. The new DaemonSet pod
can contain a single container, responsible for holding its pod network alive.
It is used as a passthrough by the CephCSI plugin pod which when mounting or
mapping will use the network namespace of this pod.

Once the pod is created get its PID and create a symlink to
`/proc/<PID>/ns/net` in the hostPath volume shared with the csi-plugin pod and
specify the path in the `netNamespaceFilePath` option.

> *Note* This Pod should have `hostPID: true` in the Pod Spec.

Reviewer 3: @Madhu-1 will it be worth mentioning the cons of this model may be as a `FIXME`, for example, unable to upgrade, binary copies issues etc?

Author: i think we dont need to worry about that one. if required as we progress will add it.

Reviewer 2: > @Madhu-1 will it be worth mentioning the cons of this model may be as a `FIXME`, for example, unable to upgrade, binary copies issues etc?

I don't think there are any upgrade issues and there is certainly no binary copying going on with this approach.  The long-running pod could literally be just a `pause` process.

Reviewer 3: AH! That sounds great and intresting too!
